[Validator] Add support of nested attributes

alexandre-daubois opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #38503
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/15541

Although the RFC (https://wiki.php.net/rfc/new_in_initializers) is in the voting phase (until 14 July), it is already well on its way to passing.

Based on @nikic's development (https://github.com/php/php-src/pull/7153), this makes the development of support possible. It will obviously take a little while before this pull request is merged.

If this pull request is OK for you, I'll get to work on writing the existing documentation for the attribute validation constraints.

![Capture d’écran du 2021-07-05 17-11-23](https://user-images.githubusercontent.com/2144837/124491886-0d2f7d80-ddb4-11eb-8147-493bdc6c48ac.png)

Not sure about the Symfony version to target, as `AtLeastOneOf` has been introduced in 5.1. Although, I couldn't find attributes validation documentation for 4.4.

alexandre-daubois committed:
[Validator] Add support of nested attributes for composite constraints

OskarStark commented:
As this is a new feature it should target `5.4` branch

alexandre-daubois commented:
Alright, I'll rebase to 5.4!

nicolas-grekas commented on src/Symfony/Component/Validator/Tests/Fixtures/Attribute/Nested/Entity.php:
```diff
@@ -0,0 +1,157 @@
+        ]),
+        Assert\Choice(choices: ['A', 'B'], message: 'Must be one of %choices%')
+    ]
+    public $firstName;
```
the Ci reports `PHP Fatal error:  Constant expression contains invalid operations`

nicolas-grekas reviewed (changes requested):
Let's wait for the RFC to be merged before merging :)

derrabus commented on src/Symfony/Component/Validator/Tests/Fixtures/Attribute/Nested/Entity.php:
```diff
@@ -0,0 +1,157 @@
+            new Assert\NotNull(),
+            new Assert\Range(min: 3),
+        ]),
+        Assert\All([
+            'constraints' => [
+                new Assert\NotNull(),
+                new Assert\Range(min: 3),
+            ],
+        ]),
```
```suggestion
        Assert\All(
            constraints: [
                new Assert\NotNull(),
                new Assert\Range(min: 3),
            ],
        ),
```

Let's switch to named arguments, like we did for the other constraints.

alexandre-daubois replied:
Looking into it!

alexandre-daubois replied:
I set `mixed` as type of constructor parameter, because of this:

https://github.com/symfony/symfony/blob/e34cd7dd2c6d0b30d24cad443b8f964daa841d71/src/Symfony/Component/Validator/Tests/Constraints/AllValidatorTest.php#L30

alexandre-daubois replied:
Didn't type-hinted the constructor parameter because of this:

https://github.com/symfony/symfony/blob/e34cd7dd2c6d0b30d24cad443b8f964daa841d71/src/Symfony/Component/Validator/Tests/Constraints/AllValidatorTest.php#L30

alexandre-daubois replied:
Done!

derrabus commented on src/Symfony/Component/Validator/Tests/Mapping/Loader/AnnotationLoaderTest.php:
```diff
@@ -185,5 +185,9 @@ public function provideNamespaces(): iterable
         }
+
+        if (\PHP_VERSION_ID >= 80100) {
+            yield 'nested_attributes' => ['Symfony\Component\Validator\Tests\Fixtures\Attribute\Nested'];
```
```suggestion
            yield 'nested_attributes' => ['Symfony\Component\Validator\Tests\Fixtures\NestedAttribute'];
```

Let's keep the directory structure flat

alexandre-daubois replied:
Alright!

alexandre-daubois commented:
Added missing tests for `Sequentially` and `AtLeastOneOf` attributes.
Added support for `Collection` as well (https://github.com/symfony/symfony/pull/41994/files#diff-d9f229d35c24533de28e1e630ee4e74afa4ce27d7af8f137d788898e0b9efc73R41).
Added documentation pull-request.

`Required` and `Optional` don't need update. I still added them in AnnotationLeaderTest.

derrabus commented:
In order to get the PHP 8.0 testsuite green, you need to exclude your new fixtures in `.github/patch-types.php`.

derrabus reviewed (approved):
LGTM, once patch-types.php has been updated.

derrabus commented on src/Symfony/Component/Validator/Tests/Mapping/Loader/AnnotationLoaderTest.php:
```diff
@@ -14,12 +14,19 @@
 use Symfony\Component\Validator\Constraints\Callback;
 use Symfony\Component\Validator\Constraints\Choice;
 use Symfony\Component\Validator\Constraints\Collection;
+use Symfony\Component\Validator\Constraints\Composite;
```
```suggestion
```

Unused import.

alexandre-daubois commented:
> In order to get the PHP 8.0 testsuite green, you need to exclude your new fixtures in `.github/patch-types.php`.

Will do, thanks!

fabpot commented:
This is now available in the 8.1 PHP branch.

fabpot commented:
Thank you @alexandre-daubois.

chekalsky commented:
@alexandre-daubois @fabpot `Optional` is still not supported in Attributes if I am not mistaken, is it done consciously? Haven't found any mentions of it.

xabbuh commented:
@chekalsky Did you experience any issues when using `Optional` or `Required` inside an attribute? As far as I can see this should work (and we also have a test case for this).

xabbuh commented on src/Symfony/Component/Validator/Tests/Fixtures/NestedAttribute/Entity.php:
```diff
@@ -0,0 +1,172 @@
+                ],
+                'bar' => new Assert\Range(min: 5),
+                'baz' => new Assert\Required([new Assert\Email()]),
+                'qux' => new Assert\Optional([new Assert\NotBlank()]),
```
@chekalsky this is the test I mean

chekalsky commented:
@xabbuh Yes, but only when it is not nested (e.g. I am trying to make one of the properties itself optional). Does it make sense?

stof commented:
Well, `Optional` and `Required` don't make sense as top-level attributes. They are meant to be used inside the `Collection` constraint.
